File: open-closed/code.py
class Discount:
    def get_discount(self, customer_type, price):
        if customer_type == "regular":
            return price * 0.9  
        elif customer_type == "vip":
            return price * 0.8  
        else:
            return price  

from abc import ABC, abstractmethod

class NewDiscount(ABC):
    @abstractmethod
    def apply_discount(self, price):
        pass

class RegularDiscount(Discount):
    def apply_discount(self, price):
        return price * 0.9  

class VIPDiscount(Discount):
    def apply_discount(self, price):
        return price * 0.8  

class SuperVIPDiscount(Discount):
    def apply_discount(self, price):
        return price * 0.7  

def calculate_discount(price, discount_strategy: Discount):
    return discount_strategy.apply_discount(price)

regular = RegularDiscount()
vip = VIPDiscount()
super_vip = SuperVIPDiscount()

print(calculate_discount(100, regular))  
print(calculate_discount(100, vip))  
print(calculate_discount(100, super_vip))  

File: open-closed/code.js
class Discount {
  getDiscount(customerType, price) {
    if (customerType === "regular") {
      return price * 0.9;
    } else if (customerType === "vip") {
      return price * 0.8;
    } else {
      return price;
    }
  }
}

const discount = new Discount();
console.log(discount.getDiscount("regular", 100)); // 90
console.log(discount.getDiscount("vip", 100)); // 80

class NewDiscount {
  applyDiscount(price) {
    return price;
  }
}

class RegularDiscount extends Discount {
  applyDiscount(price) {
    return price * 0.9;
  }
}

class VIPDiscount extends Discount {
  applyDiscount(price) {
    return price * 0.8;
  }
}

class SuperVIPDiscount extends Discount {
  applyDiscount(price) {
    return price * 0.7;
  }
}

function calculateDiscount(price, discountStrategy) {
  return discountStrategy.applyDiscount(price);
}

console.log(calculateDiscount(100, new RegularDiscount()));
console.log(calculateDiscount(100, new VIPDiscount()));
console.log(calculateDiscount(100, new SuperVIPDiscount()));
